nav2_motion_primitives/include/nav2_motion_primitives/motion_primitive.hpp

```diff
@@ -24,127 +24,137 @@
-  virtual nav2_tasks::TaskStatus onRun(
-    const typename /*nav2_tasks::*/ CommandMsg::SharedPtr command) = 0;
+  virtual Status onRun(const std::shared_ptr<const typename ActionT::Goal> command) = 0;
+
 
   // This is the method derived classes should mainly implement
   // and will be called cyclically while it returns RUNNING.
   // Implement the behavior such that it runs some unit of work on each call
   // and provides a status.
-  virtual nav2_tasks::TaskStatus onCycleUpdate(
-    /*nav2_tasks::*/ ResultMsg & result) = 0;
+  virtual Status onCycleUpdate() = 0;
 
-  // Runs the behavior
-  nav2_tasks::TaskStatus run(
-    const typename /*nav2_tasks::*/ CommandMsg::SharedPtr command)
-  {
-    RCLCPP_INFO(node_->get_logger(), "%s attempting behavior", taskName_.c_str());
+protected:
+  rclcpp::Node::SharedPtr node_;
+  std::string primitive_name_;
+  std::shared_ptr<nav2_robot::Robot> robot_;
+  std::unique_ptr<ActionServer> action_server_;
 
-    auto status = onRun(command);
+  void configure()
+  {
+    RCLCPP_INFO(node_->get_logger(), "Configuring ", primitive_name_.c_str());
 
-    if (status != nav2_tasks::TaskStatus::SUCCEEDED) {
-      return nav2_tasks::TaskStatus::FAILED;
-    }
+    robot_ = std::make_unique<nav2_robot::Robot>(
+      node_->get_node_base_interface(),
+      node_->get_node_topics_interface(),
+      node_->get_node_logging_interface(),
+      true);
 
-    ResultMsg result;
-    status = cycle(result);
-    task_server_->setResult(result);
+    action_server_ = std::make_unique<ActionServer>(node_, primitive_name_,
+        std::bind(&MotionPrimitive::execute, this, std::placeholders::_1));
+  }
 
-    return status;
+  void cleanup()
+  {
+    robot_.reset();
+    action_server_.reset();
   }
 
-protected:
-  nav2_tasks::TaskStatus cycle(ResultMsg & result)
+  void execute(const typename std::shared_ptr<GoalHandle> goal_handle)
   {
-    auto time_since_msg = std::chrono::system_clock::now();
-    auto start_time = std::chrono::system_clock::now();
-    auto current_time = std::chrono::system_clock::now();
+    RCLCPP_INFO(node_->get_logger(), "Attempting ", primitive_name_.c_str());
+
+    initialCheck(goal_handle);
+
+    // Log a message every second
+    auto timer = node_->create_wall_timer(1s,
+        [&]() {RCLCPP_INFO(node_->get_logger(), "%s running...", primitive_name_.c_str());});
 
-    auto status = nav2_tasks::TaskStatus::FAILED;
+    loop(goal_handle);
 
+    stopRobot();
+  }
+
+  void initialCheck(const typename std::shared_ptr<GoalHandle> goal_handle)
+  {
+    if (onRun(goal_handle->get_goal()) != Status::SUCCEEDED) {
+      RCLCPP_INFO(node_->get_logger(), "Initial checks failed for ", primitive_name_.c_str());
+      goal_handle->abort(std::make_shared<typename ActionT::Result>());
+    }
+  }
+
+  void loop(const typename std::shared_ptr<GoalHandle> goal_handle)
+  {
+    auto result = std::make_shared<typename ActionT::Result>();
     rclcpp::Rate loop_rate(10);
```

Comment:
You've hard coded 10hz for all potential recoveries, shouldnt this be configurable? Some recoveries may need to run faster or slower than others.

Reply:
Yes, possibly. For the current primitives, the hardcoded rate is good enough, I guess once we find a recovery that needs a different rate and understand the use case we can modify it.

Reply:
Ok, I just filed a ticket so its not forgotten. I might just do it after this diff lands as a future proofing method.